### PR TITLE
Optional actions in operation

### PR DESCRIPTION
**Pull request by fenris85**
Optional actions in operation are marked by passing additional array when creating operation in the OperationRegistry, e.g.:

`[ Action1, Action2, Action3, Actions4]` - actions hashes list
`[false, true, false, false]` - optionality list

In above case `Action2` can be skipped. 

Skipping when forming an operation on FE can be achieved by adding param `skipped = true`. This action still has to exist in operation, but all params can be omitted. It requires only actionHash and `skipped` param.

### CODE REVIEW

File: contracts/core/OperationsRegistry.sol [hunk 1]
```diff
@@ -5,12 +5,14 @@ import { OPERATIONS_REGISTRY } from "./constants/Common.sol";
 
 struct StoredOperation {
   bytes32[] actions;
+  bool[] optional;
   string name;
 }
```
Comment (peculiarity) [inline contracts/core/OperationsRegistry.sol:12]:
```
struct Action {
  hash: bytes32,
  isSKippable: bool,
 }
 ```

Comment (peculiarity) [reply contracts/core/OperationsRegistry.sol:12]:
```
struct Operation {
   actions: Action[],
   string: name
}
```

File: contracts/core/OperationsRegistry.sol [hunk 2]
```diff
@@ -43,21 +45,24 @@ contract OperationsRegistry {
    * @notice Adds an Operation's Actions keyed to a an operation name
    * @param name The Operation name
    * @param actions An array the Actions the Operation consists of
+   * @param optional An array indicating given Action can be skipped
    */
-  function addOperation(string memory name, bytes32[] memory actions) external onlyOwner {
-    operations[name] = StoredOperation(actions, name);
+  function addOperation(string memory name, bytes32[] memory actions, bool[] memory optional) external onlyOwner {
```
Comment (peculiarity) [inline contracts/core/OperationsRegistry.sol]: ```addOperation(Operation calldata operation) external onlyOwner```

File: test/common/OptionalActions.test.ts [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+import { executeThroughProxy } from '../../helpers/deploy'
+import { restoreSnapshot } from '../../helpers/restoreSnapshot'
+import { ServiceRegistry } from '../../helpers/serviceRegistry'
+import { RuntimeConfig } from '../../helpers/types/common'
+import { OperationsRegistry } from '../../helpers/wrappers/operationsRegistry'
+import { testBlockNumber } from '../config'
+import { DeployedSystemInfo } from '../deploySystem'
+import { initialiseConfig } from '../fixtures/setup'
+
+const createAction = ActionFactory.create
```
Comment (peculiarity) [inline test/common/OptionalActions.test.ts:19]: IMO tests are not written well. 
Let's for example comment out the call

```
// if(!calls[current].skipped) {
//   address target = registry.getServiceAddress(calls[current].targetHash);
//   target.execute(calls[current].callData);
// }
 ```
 Which is in the `OperationExecutor`. These test will pass. 
 I'm not sure what exactly is being test. Also you shouldn't be testing a given contract through proxy calls ( I mean one contract, calling the other contract ) and verifying if the TX is successful.

Comment (peculiarity) [reply test/common/OptionalActions.test.ts:19]: Write tests that given `addOperation` and some params ` `getOperation` will return same values.
That will be under `OperationsRegistry.test.ts`
We should have `OperationsExecutor` which must contain logic about testing the login within `executeOp`. 
You can call an action for example that will fire an event and check if the TX contains that event or if state is changed or smth like that. 
We have only integration tests which is not good.

**Review by peculiarity — CHANGES_REQUESTED** [review]

File: contracts/core/OperationsRegistry.sol (hunk 1 above)
Comment (zerotucks) [reply contracts/core/OperationsRegistry.sol:12]: Kuba mentioned he's already looked into this. And dynamic arrays of structs create issues elsewhere

File: test/common/OptionalActions.test.ts (hunk 3 above)
Comment (zerotucks) [reply test/common/OptionalActions.test.ts:19]: Broadly agree. I'm going to ask Frank if we can have a ticket for this next sprint.

**Review by zerotucks — APPROVED** [review]

**Pull request merged**